### DISCUSSION TITLE
Can we get vuln-type filtering for licenses as well?

### DISCUSSION THREAD

**Discussion by mastacheata** [aquasecurity/trivy#4991] [Ideas]
### Description

Hey there,

it would be great if the license scanner results could be filtered by type before outputting them.

I think the vuln-type filter would be especially useful to filter out OS packages.
The typical use case for trivy is not people building new Linux distributions and sharing them publicly, but people building and sharing applications that just so happen to run on a Linux system. In the later case, the operating system packages being GPL licensed is of no concern at all, but a library-type dependency with a viral license is extremely important to consider.

So far, the only option around that is to filter the results by specifying a custom template or filtering the json or text outputs using third party tools.

Thanks for considering this.

### Target

Container Image

### Scanner

License

**Comment by knqyf263** [comment]
Thanks for your idea. It makes sense. `--vuln-type` was added a long time ago when Trivy had only a vulnerability scanner. Maybe `--pkg-type` is better now?

**Reply by mastacheata** [reply]
Maybe Target is the better wording? In the documentation you already refer to the vuln-type switch as filtering for Vulnerability Target, so generalizing that to just target sounds reasonable.

I don't have any strong opinions on the name of thing, though.

**Reply by knqyf263** [reply]
We already use the "Target" term. But thanks for your suggestion.
https://aquasecurity.github.io/trivy/v0.44/docs/

**Reply by mastacheata** [reply]
Any progress on this?
If you give me some hints on where to look I'd be happy to look into this myself, but I have a really hard time understanding the structure of trivy and what happens where.

**Reply by dus7eh** [reply]
That would be indeed useful. Right now we can do of course some post processing of json reports but this way we loose the ability to get proper exit code when library package of an unwanted severity is detected.

**Reply by mastacheata** [reply]
My current workaround is to add this rego file:
```rego
package trivy

default ignore = false

ignore {
    input.FilePath == ""
}
```

This works because I found that OS packages don't have a FilePath reference, while all other findings have a file reference (vulnerabilities, misconfigs, licenses, secrets)
Obviously this is bound to break and it would be better suited to write a module to filter out the OS packages from the licenses, but I'm not confident enough in Go to write and publish something like that with the sparse documentation there currently is for how to write a module.

**Comment by mastacheata** [comment]
I think this patch should solve that problem. (https://github.com/aquasecurity/trivy/compare/main...mastacheata:trivy:feat/observe-pkg-types-option-in-license-scanner)
Unsure if this would be considered minor enough to open a PR directly or if this is a big enough change to warrant having an issue opened by the maintainers first.

I'm not 100% sure which (if any) of the two cases the "Scan licenses in file headers or license files" case should fall so I've left that enabled in either case for now, but I'm happy to reconsider that if there's input.

Also the Code-Style within that function appears inconsistent to me: For the OS Packages and the File packages you assign the result to a variable first and only append them to the results set if they're not empty, but for the language-specific files you use a different syntax.

**Reply by DmitriyLewen** [reply]
Hello @mastacheata 
It looks as good solution.
Can you open PR for these changes?

**Reply by mastacheata** [reply]
Done:
https://github.com/aquasecurity/trivy/pull/9091

Added an entry to the License Documentation (also blatantly copied from the vulnerability docs), I don't have any tests, though.

**Comment by DmitriyLewen** [comment]
track #9103
